pkg/jaegerquery/trace_query.go

```diff
@@ -0,0 +1,450 @@
+	LEFT JOIN
+		_ps_trace.instrumentation_lib inst_lib ON s.instrumentation_lib_id = inst_lib.id
+	LEFT JOIN
+		_ps_trace.schema_url inst_lib_url ON inst_lib_url.id = inst_lib.schema_url_id
+	LEFT JOIN
+		_ps_trace.link lk ON lk.trace_id = s.trace_id AND lk.span_id = s.span_id
+	WHERE
+	  %s
+	GROUP BY
+	  s.trace_id,
+	  s.span_id,
+	  s.parent_span_id,
+	  s.start_time,
+	  s.end_time,
+	  s.resource_tags,
+	  s.span_tags,
+	  sn.name,
+	  s_url.url,
+	  inst_lib.name,
+	  inst_lib.version,
+	  inst_lib_url.url
+	%s`
+
+	subqueryFormat = `
+	SELECT
+		DISTINCT trace_id
+	FROM _ps_trace.span s
+	WHERE
+		%s
+	`
+)
+
+func buildTraceQuery(traceIDClause string, multipleTraces bool) string {
+	orderBy := ""
+	if multipleTraces {
+		orderBy = "ORDER BY s.trace_id"
+	}
+	return fmt.Sprintf(
+		fullTraceSQLFormat,
+		traceIDClause, orderBy)
+}
+
+func findTracesQuery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	subquery, params := buildTraceIDSubquery(q)
+	traceIDClause := fmt.Sprintf("s.trace_id IN (%s)", subquery)
+	return buildTraceQuery(traceIDClause, true), params
+}
+
+func findTraceIDsQuery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	subquery, params := buildTraceIDSubquery(q)
+	return subquery, params
+}
+
+func getTraceQuery(traceID model.TraceID) (string, []interface{}, error) {
+	var b [16]byte
+	n, err := traceID.MarshalTo(b[:])
+	if n != 16 || err != nil {
+		return "", nil, fmt.Errorf("marshaling TraceID: %w", err)
+	}
+
+	var uuid pgtype.UUID
+	uuid.Set(b)
+	params := []interface{}{uuid}
+
+	traceIDClause := "s.trace_id = $1"
+	return buildTraceQuery(traceIDClause, false), params, nil
+}
+
+func buildTraceIDSubquery(q *spanstore.TraceQueryParameters) (string, []interface{}) {
+	quals := make([]string, 0, 15)
+	params := make([]interface{}, 0, 15)
+
+	if len(q.ServiceName) > 0 {
+		params = append(params, q.ServiceName)
+		qual := fmt.Sprintf(`s.resource_tags OPERATOR(ps_trace.?) ('service.name' OPERATOR(ps_trace.==) $%d)`, len(params))
+		quals = append(quals, qual)
+	}
+	if len(q.OperationName) > 0 {
+		params = append(params, q.OperationName)
+		qual := fmt.Sprintf(`s.name_id =
+		(
+			SELECT
+				id
+			FROM
+				_ps_trace.span_name sn
+			WHERE
+				name = $%d
+		)`, len(params))
+		quals = append(quals, qual)
+	}
+	if len(q.Tags) > 0 {
+		for k, v := range q.Tags {
+			params = append(params, k, v)
+			qual := fmt.Sprintf(`s.span_tags OPERATOR(ps_trace.?) ($%d OPERATOR(ps_trace.==) $%d)`, len(params)-1, len(params))
+			quals = append(quals, qual)
+		}
+	}
+
+	//todo check the inclusive semantics here
+	var defaultTime time.Time
+	if q.StartTimeMin != defaultTime {
+		params = append(params, q.StartTimeMin)
+		quals = append(quals, fmt.Sprintf(`s.start_time >= $%d`, len(params)))
+	}
+	if q.StartTimeMax != defaultTime {
+		params = append(params, q.StartTimeMax)
+		quals = append(quals, fmt.Sprintf(`s.start_time <= $%d`, len(params)))
+	}
```

Comment:
I recall from Matous's tests, using `between` was more performant than using `AND`. Same goes for duration below.